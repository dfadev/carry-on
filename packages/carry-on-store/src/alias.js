import { forceArray, isString } from "carry-on-utils";
import { register } from "./store";

export const state = (init, id, path, dispose) => {
  // storeId can be the first parameter
  if (isString(init)) {
    const actualId = init;
    init = id;
    id = actualId;
  }

  const inits = forceArray(init).map(item => ({
    state: item,
    dispose
  }));

  register(inits, id, path);
};

export const middleware = (init, id, path, dispose) => {
  // storeId can be the first parameter
  if (isString(init)) {
    const actualId = init;
    init = id;
    id = actualId;
  }

  const inits = forceArray(init).map(item => ({
    middleware: item,
    dispose
  }));

  register(inits, id, path);
};
