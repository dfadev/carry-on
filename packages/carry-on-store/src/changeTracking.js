/** @format **/
import { keys, proxyState, deproxify, mutateSetA } from "carry-on-utils";

export function compareChanges(changes, affected) {
  const queue = [];
  queue.push({ changes, affected });

  while (queue.length > 0) {
    const item = queue.pop();
    const entries = item.changes;

    for (let i = 0, len = entries.length; i < len; i++) {
      const entry = entries[i];
      const key = entry.key;
      const affectedValue = item.affected[key];

      if (affectedValue === true) return true;
      if (affectedValue !== undefined) {
        const nextChanges = entry.changes;

        if (nextChanges === true) return true;
        queue.push({
          changes: nextChanges,
          affected: affectedValue
        });
      }
    }
  }

  return false;
}

export function trackChanges(state, select) {
  const trappedState = proxyState(state);
  const selectedState = select(trappedState.state);
  trappedState.seal();
  const affected = trappedState.affected;
  const deproxified = deproxify(selectedState);
  const finalState = deproxified !== undefined ? deproxified : selectedState;

  return {
    finalState,
    affected
  };
}

export function calculateChanges(patches) {
  const stage1 = {};
  for (let i = 0, len = patches.length; i < len; i++)
    mutateSetA(stage1, patches[i].path, true);

  // precompute object walk so keys is called the least amount necessary
  const stage2 = [];
  const queue = [];
  queue.push({ keys: keys(stage1), changes: stage1, out: stage2 });

  while (queue.length > 0) {
    const item = queue.pop();

    for (let i = 0, len = item.keys.length; i < len; i++) {
      const key = item.keys[i];
      const changes = item.changes[key];
      if (changes === true) {
        item.out.push({ key, changes });
      } else {
        const nextChanges = [];
        queue.push({
          keys: keys(changes),
          changes,
          out: nextChanges
        });
        item.out.push({ key, changes: nextChanges });
      }
    }
  }

  return stage2;
}
