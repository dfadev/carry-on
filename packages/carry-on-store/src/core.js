/** @format **/
import { isFunction } from "./utils";

export default function makeStoreModule(defaultId, extra = () => ({})) {
  // wrap a function with middleware
  const applyMiddleware = (middlewares, fn, apply) => {
    if (!Array.isArray(middlewares)) middlewares = [middlewares];
    for (const middleware of middlewares) fn = apply(middleware, fn);
    return (...args) => fn(...args);
  };

  // middleware initialize message type
  const initMessage = "Initialize";

  const createPlugins = (store, plugins) => {
    if (!plugins) return;

    // create plugins
    store.plug = {};

    if (!Array.isArray(plugins)) plugins = [plugins];

    for (const plugin of plugins) {
      const { id, state, dispatch } = plugin;
      if (state)
        store.plug[id] = isFunction(state)
          ? state(store.dispatch, store.query, store.plug)
          : state;
      if (dispatch)
        store.d = applyMiddleware(dispatch, store.d, (middleware, fn) =>
          middleware(fn, store.query, store.plug)
        );
    }
  };

  // a map of stores
  const stores = {};

  // create a store
  function create(id) {
    return Object.assign({ id, pending: [] }, extra());
  }

  // delete a store
  function deleteStore(id = defaultId) {
    delete stores[id];
  }

  // lookup a store
  const useStore = (id = defaultId) => stores[id] || (stores[id] = create(id));

  // register state
  const register = (init, id = defaultId) => {
    const { dispatch, query, pending } = useStore(id);
    // queue if no dispatch available yet
    if (dispatch)
      dispatch(state =>
        Object.assign(state, isFunction(init) ? init(dispatch, query) : init)
      );
    else pending.push(init);
  };

  const connect = ({ id, producer, publish, plugins, init }) => {
    const store = useStore(id);
    if (store.dispatch) throw new Error("Already connected");

    // the producer creates new states from old states by executing actions
    store.producer = producer || ((state, action) => ({ ...action(state) }));

    // query provides a copy of state created by the producer
    store.query = (action = identity => identity, ...args) =>
      store.producer(store.state, action, ...args);

    // run producer action and set state
    store.d = (action, type, force) => {
      const nextState = (store.state = force
        ? action(store.state)
        : store.producer(store.state, action));
      if (type !== initMessage) {
        publish(nextState);
      }
      return nextState;
    };
    store.dispatch = (...args) => store.d(...args);

    createPlugins(store, plugins);
    store.dispatch = store.d;

    // create initial state
    store.state = Object.assign(
      {},
      // execute store initialization
      isFunction(init) ? init(store.dispatch, store.query, store.plug) : init,
      // execute pending store initialization
      ...store.pending.map(pending =>
        pending(store.dispatch, store.query, store.plug)
      ),
      // add standard keys
      {
        id: store.id,
        dispatch: store.dispatch,
        query: store.query,
        plug: store.plug
      }
    );

    // initialize middleware with state
    if (plugins) store.dispatch(() => store.state, initMessage);

    return store.state;
  };

  return {
    useStore,
    deleteStore,
    connect,
    defaultId,
    register
  };
}
