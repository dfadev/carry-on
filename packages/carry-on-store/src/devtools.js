/** @format **/
export default function devTools({ timeTravel = true } = {}) {
  // dev tools connections
  const connections = [],
    subscriptions = [],
    // time travel state tracking (only for immutable state)
    time = {},
    // check for dev tools extension
    devToolsExt = window && window.__REDUX_DEVTOOLS_EXTENSION__;

  if (!devToolsExt) return {};

  return {
    middleware: ({ dispatch, id }) =>
      function devToolsMiddleware(action, type = "Dispatch", ...args) {
        const state = dispatch(action, type, ...args);

        // exit when no action
        if (!action) return state;
        const name = id;

        let connection = connections[name];
        if (!connection)
          connection = connections[name] = devToolsExt.connect({ name });

        // support time traveling
        if (timeTravel) {
          const states = time[name] || (time[name] = []);
          states.push(state);
          subscriptions[name] ||
            (subscriptions[name] = connection.subscribe(
              msg =>
                msg.type === "DISPATCH" &&
                msg.payload &&
                msg.payload.type === "JUMP_TO_STATE" &&
                dispatch(s => {
                  if (s === undefined) s = {};
                  const keys = Object.keys(s);
                  for (let i = 0, len = keys.length; i < len; i++)
                    delete s[keys[i]];
                  return Object.assign(s, states[msg.payload.index]);
                }, "Time Travel")
            ));
        }

        // send devtools an state update message
        connection.send({ type }, state);

        return state;
      }
  };
}
