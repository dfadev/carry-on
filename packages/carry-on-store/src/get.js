import { isString } from "carry-on-utils";
import { getStore } from "./store";

export default function get(fn, id, opts) {
  if (isString(fn)) {
    const actualId = fn;
    fn = id;
    id = actualId;
  }

  const store = getStore(id);
  return store.get(fn, opts);
}
