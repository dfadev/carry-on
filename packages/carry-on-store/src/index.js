/** @format **/
export default function makeStoreModule(defaultId, extra = () => ({})) {
  const isFunction = thing =>
    !!(thing && thing.constructor && thing.call && thing.apply);

  // wrap a function with middleware
  const applyMiddleware = (middlewares, fn, apply) => {
    if (!Array.isArray(middlewares)) middlewares = [middlewares];
    for (const middleware of middlewares) fn = apply(middleware, fn);
    return (...args) => fn(...args);
  };

  // middleware initialize message type
  const initMessage = "Initialize";

  const createPlugins = (store, plugins) => {
    // create plugins
    store.plug = {};

    if (!Array.isArray(plugins)) plugins = [plugins];

    // dispatch
    for (const plugin of plugins) {
      const { dispatch } = plugin;
      if (dispatch)
        store.d = applyMiddleware(dispatch, store.d, (middleware, fn) =>
          middleware(fn, store.query, store.plug)
        );
    }

    // state
    for (const plugin of plugins) {
      const { id, state } = plugin;
      if (state)
        store.plug[id] = isFunction(state)
          ? state(store.d, store.query, store.plug)
          : state;
    }
  };

  // a map of stores
  const stores = {};

  // create a store
  function create(id) {
    return Object.assign({ id, pending: [] }, extra());
  }

  // delete a store
  function deleteStore(id = defaultId) {
    delete stores[id];
  }

  // lookup a store
  const useStore = (id = defaultId) => stores[id] || (stores[id] = create(id));

  // register state
  const register = (init, id = defaultId) => {
    const { dispatch, query, pending } = useStore(id);
    // queue if no dispatch available yet
    if (dispatch)
      dispatch(state =>
        Object.assign(
          {},
          state,
          isFunction(init) ? init(dispatch, query) : init
        )
      );
    else pending.push(init);
  };

  const connect = ({ id, producer, publish, plugins, init } = {}) => {
    const store = useStore(id);
    if (store.dispatch) throw new Error("Already connected");

    // the producer creates new states from old states by executing actions
    store.producer = producer || ((state, action) => action({ ...state }));

    // query provides a copy of state created by the producer
    store.query = (action = identity => identity, ...args) =>
      store.producer(store.state, action, ...args);

    // run producer action and set state
    store.d = (action, type, force) => {
      const nextState = (store.state = force
        ? action(store.state)
        : store.producer(store.state, action));
      if (publish && type !== initMessage) {
        publish(nextState);
      }
      return nextState;
    };

    if (plugins) createPlugins(store, plugins);
    store.dispatch = store.d;

    // create initial state
    store.state = Object.assign(
      {},
      // execute store initialization
      isFunction(init) ? init(store.dispatch, store.query, store.plug) : init,
      // execute pending store initialization
      ...store.pending.map(
        pending =>
          isFunction(pending)
            ? pending(store.dispatch, store.query, store.plug)
            : pending
      ),
      // add standard keys
      {
        dispatch: store.dispatch,
        query: store.query,
        plug: store.plug
      },
      store.id ? { id: store.id } : undefined
    );

    // initialize middleware with state
    if (plugins) store.dispatch(() => store.state, initMessage);

    return store.state;
  };

  return {
    useStore,
    deleteStore,
    connect,
    defaultId,
    register
  };
}
