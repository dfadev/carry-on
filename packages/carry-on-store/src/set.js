/** @format **/
import { isString } from "carry-on-utils";
import { getStore } from "./store";

export default function set(fn, id) {
  if (isString(fn)) {
    const actualId = fn;
    fn = id;
    id = actualId;
  }

  const store = getStore(id);
  if (!store) throw new Error("store does not exist");
  if (!store.set) throw new Error("store not connected");
  store.set(fn);
}
