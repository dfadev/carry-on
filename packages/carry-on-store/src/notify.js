import { compareChanges } from "./changeTracking";

export default function notify() {
  const subscribers = new Map();

  function subscribe(fn, watch) {
    subscribers.set(fn, watch);
    return function unsusbscribe() {
      subscribers.delete(fn);
    };
  }

  const plugin = {
    middleware: ({ next, getChanges, wrap, isNested }) =>
      function notifyMiddleware(action, type, opts, ...args) {
        const state = next(action, type, opts, ...args);
        if (isNested()) return state; // don't notify on nested sets

        const changes = getChanges && getChanges();

        const notifySubs = () => {
          const notifyThese = [];
          const entries = subscribers.entries();
          let entry = entries.next();
          while (!entry.done) {
            const [fn, watch] = entry.value;
            if (watch === undefined) notifyThese.push(fn);
            else {
              const hasChanges = compareChanges(changes, watch);
              if (hasChanges) notifyThese.push(fn);
            }
            entry = entries.next();
          }

          for (let i = 0, len = notifyThese.length; i < len; i += 1)
            notifyThese[i](state, changes);
        };

        if (opts && opts.immediate) notifySubs();
        else wrap(notifySubs);

        return state;
      },
    dispose: () => {
      subscribers.clear();
    }
  };

  return {
    subscribers,
    subscribe,
    plugin
  };
}
