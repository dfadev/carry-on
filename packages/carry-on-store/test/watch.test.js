/**
 * @jest-environment jsdom
 */
import { getStore, deleteStore, connect, register, watch, Watch } from "../src";

test("watch", () => {
  Watch.Debug = true;
  Watch.Verbose = true;

  connect();
  let fnCalled = 0;
  const fn = (state, changes) => {
    if (state.a.field) fnCalled++;
  };
  let elseCalled = 0;
  const elseFn = state => {
    elseCalled += 1;
  };
  let errorCalled = 0;
  const errorFn = (error, state) => {
    errorCalled += 1;
  };

  register({ state: { a: { field: "value" } } });
  const store = getStore();

  const unsub_ws1 = watch({
    id: "WS1",
    register: { state: { junk: "value" } },
    if: fn
  });
  const unsub_ws2 = watch({
    id: "WS2",
    if: s => s.a.field > 1,
    then: field => field && fnCalled++,
    else: elseFn,
    throttle: 1000
  });

  const unsub_ws3 = watch({
    debug: false,
    id: "WS3",
    path: "a.field",
    def: 0,
    if: v => v > 1,
    then: () => {
      fnCalled += 1;
      throw new Error("fake error");
    },
    else: elseFn,
    error: errorFn
  });

  store.set(state => {
    state.b = 2;
  });

  store.set(state => {
    state.a.field = 2;
  });

  store.set(state => {
    state.c = 2;
  });

  unsub_ws1();
  unsub_ws2();
  unsub_ws3();

  store.set(state => {
    state.a.field = 3;
  });

  expect(fnCalled).toBe(4);
  expect(elseCalled).toBe(2);
  expect(errorCalled).toBe(1);

  expect(store.get()).toMatchSnapshot();

  deleteStore();
});
