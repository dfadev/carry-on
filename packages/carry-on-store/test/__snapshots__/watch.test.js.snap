// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`watch 1`] = `
Object {
  "a": Object {
    "field": 3,
  },
  "b": 2,
  "c": 2,
  "junk": "value",
}
`;
